### DISCUSSION TITLE
A schema for an array that includes at least one item of a certain type

### DISCUSSION THREAD

**Discussion by vincent822** [colinhacks/zod#2937] [Q&A]
I imagine something like:

```ts
const schema = z.any().array().includes(z.string())
```

Seems like a very common use case (I need it right now for my project).
Is there a way to do such thing without refine?

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const schema = z.string().array().min( 1 )

console.log( schema.parse( [ 'foo' ] ) ) // [ 'foo' ]

const result = schema.safeParse( [] )
!result.success && console.log( result.error.issues )
// [
//     {
//         code: 'too_small',
//         minimum: 1,
//         type: 'array',
//         inclusive: true,
//         exact: false,
//         message: 'Array must contain at least 1 element(s)',
//         path: []
//     }
// ]
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by JacobWeisenburger** [comment]
I'm assuming you changed the title because I didn't give you the answer you were looking for. Could you please provide some data that would pass and some that would fail the schema you are wanting to make?

I would love to help, but I need a little more info to be able to help.

**Comment by vincent822** [comment]
Hi, thank you for your answer.
I wasn't precise enough in my description.

I need it to be an array that can contain many types (or many schemas)
For example, I want to make sure an array contains at least one item of kind `bookSchema` and at least one item of kind `movieSchema`

```
let's say I have 3 schemas:
bookSchema = z.object(...)
movieSchema = z.object(...)
comicsSchema = z.object(...)

const validData = [ book1, book2, movie1]
const invalidData = [book1, comics1]

// This is the feature I want:
const mySchema = z.any().array().includes(bookSchema).includes(movieSchema)

// And the result:
mySchema.parse(validData) // pass
mySchema.parse(invalidData) // fail -  because no item in the list passes validation for movieSchema 

```

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const bookSchema = z.object( { type: z.literal( 'book' ) } )
const movieSchema = z.object( { type: z.literal( 'movie' ) } )
const comicSchema = z.object( { type: z.literal( 'comic' ) } )

const schema = z.object( {
    books: bookSchema.array().min( 1 ),
    movies: movieSchema.array().min( 1 ),
    comics: comicSchema.array(),
} )

const validData = {
    books: [ { type: 'book' } ],
    movies: [ { type: 'movie' } ],
    comics: [ { type: 'comic' } ]
}
console.log( schema.safeParse( validData ).success )

const invalidData = {
    books: [ { type: 'book' } ],
    movies: [],
    comics: [ { type: 'comic' } ]
}
console.log( schema.safeParse( invalidData ).success )
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by vincent822** [comment]
No, because I have an array of products, not object.

 Tomorrow I may have a new kind of product which adheres to a new schema
(e.g. toySchema) and I would want to validate whether a product list
contains one (or more) items which are toys and one or more items which are
books.

According to your suggestion, I would have to keep adding fields to the
base schema whenever I have a new specific product schema.

**Reply by JacobWeisenburger** [reply]
as far as I can see, there isn't a built in way to do what you are asking, so you will likely need to do something with `refine` or `transform`.

**Comment by vincent822** [comment]
Yes, that’s why I submitted that as an issue first, as a feature request.

**Reply by JacobWeisenburger** [reply]
> ... (I need it right now for my project).
> Is there a way to do such thing without refine?

Can you see how this might seem to someone else that you are looking for a workaround because you don't have time to wait for it to be implemented as a core feature?

**Reply by vincent822** [reply]
Yes, maybe I wasn't very clear.
I can easily implement it right now with refine but I thought it would be much more nice if such method exists because validating whether an array of multiple types contains an item of specific type seems like a common use (at least in my project it is).

**Reply by JacobWeisenburger** [reply]
Honestly, I have never seen anyone try to do what you are doing. So if it were me, I would figure out a way to not combine different types of data into an array. Arrays are easiest when they only have 1 type of data. Is that something you are able to change? or is the project bigger than just you?

**Reply by lsarrazi** [reply]
@JacobWeisenburger your comment miss the issue entirely and falsely assumes types of data have to be different for this feature to be relevant
Plus what people most likely do is unsufficient to exhaust what should be in library, use case covering is

**Comment by vincent822** [comment]
It’s bigger than just me…
But I also don’t think it such a weird structuring.

Example:
You have shopping carts which are arrays of products (it makes no sense to
define them as objects). Then, for example, you want to define a
discountEntitledCartScema as a cart that contains both a book and a movie.
How would you do it differently?

**Reply by JacobWeisenburger** [reply]
if you must combine different types into an array, then it's probs a good idea to have a `type` field or something to easily tell them apart. Do you have such a field?

**Reply by yandi-banyu** [reply]
I do have such field, and how should I go about doing it without refine? In my case, I have a base object and the extended type can be typed as discriminated union. However, I want an array of the discriminated union to contain at least one type of the discriminated union.

**Comment by lsarrazi** [comment]
Here's a workaround implementation since nobody is taking that seriously apparently

```ts
import { ZodArray, ZodIssue, ZodTypeAny, ZodEffects } from 'zod';
declare module 'zod' {
    interface ZodArray<T extends ZodTypeAny> {
        /** This function requires you have listed the existing fields in the
         * array schema before calling it.
         * 
         * @example
         * object({ a: string().optional() }).array().includes(object({ a: literal('hey') }))
         * // This won't work:
         * object({}).array().includes(object({ a: literal('hey') }))
        */
        includes(schema: T, message?: string): ZodEffects<ZodArray<T>>;

        /** This function requires you have listed the existing fields in the
         * array schema before calling it.
         * 
         * @example
         * object({ a: string().optional() }).array().excludes(object({ a: literal('hey') }))
         * // This won't work:
         * object({}).array().excludes(object({ a: literal('hey') }))
        */
        excludes(schema: ZodTypeAny, message?: string): ZodEffects<ZodArray<T>>;
    }

    interface ZodEffects<T extends ZodTypeAny> {
        /** This function requires you have listed the existing fields in the
         * array schema before calling it.
         * 
         * @example
         * object({ a: string().optional() }).array().includes(object({ a: literal('hey') }))
         * // This won't work:
         * object({}).array().includes(object({ a: literal('hey') }))
        */
        includes(schema: ZodTypeAny, message?: string): ZodEffects<T>;

        /** This function requires you have listed the existing fields in the
         * array schema before calling it.
         * 
         * @example
         * object({ a: string().optional() }).array().excludes(object({ a: literal('hey') }))
         * // This won't work:
         * object({}).array().excludes(object({ a: literal('hey') }))
        */
        excludes(schema: ZodTypeAny, message?: string): ZodEffects<T>;
    }
}

ZodEffects.prototype.includes =
    ZodArray.prototype.includes = function (schema: ZodTypeAny, message?: string) {
        return this.superRefine((arr, ctx) => {
            const issues = new Map<string, ZodIssue>();

            if (!Array.isArray(arr)) {
                ctx.addIssue({
                    code: 'custom',
                    message: 'The value is not an array. Includes cannot be applied.\n',
                });
                return false;
            }

            const success = arr.some(item => {
                const result = schema.safeParse(item);

                if (!result.success) {
                    for (const issue of result.error.issues) {
                        issues.set(issue.path.join('.') + ':' + issue.code, issue);
                    }
                }
                return result.success;
            });

            if (!success)
                [...issues.values()].forEach(issue => ctx.addIssue(issue));

            return success;
        }, {
            message: message || 'The array must contain at least one item matching the specified schema.\n',
        }) as unknown as any;
    };

ZodEffects.prototype.excludes =
    ZodArray.prototype.excludes = function (schema: ZodTypeAny, message?: string) {
        return this.superRefine((arr, ctx) => {

            if (!Array.isArray(arr)) {
                ctx.addIssue({
                    code: 'custom',
                    message: 'The value is not an array. Excludes cannot be applied.\n',
                });
                return false;
            }

            return arr.every(item => {
                const result = schema.safeParse(item);

                if (result.success) {
                    ctx.addIssue({
                        code: 'custom',
                        message: message || 'The array must not contain any item matching the specified schema.\n',
                    });
                }
                return !result.success;
            });
        }, {
            message: message || 'The array must not contain any item matching the specified schema.\n',
        }) as unknown as any;
    };
    ```
